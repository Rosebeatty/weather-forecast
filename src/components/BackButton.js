import { Link } from "react-router-dom";
import styled from "styled-components";

const Back = styled(Link)`
  position: absolute;
  top: 2%;
  left: 5%;
`;

// Button back to dashboard
export const BackButton = () => (
  <Back id="back" to="/">
    <img
      style={{ width: "40px" }}
      alt="back icon"
      src="https://img.icons8.com/material-sharp/96/000000/back--v1.png"
    />
  </Back>
);
