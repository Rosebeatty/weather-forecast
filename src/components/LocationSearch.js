import React, { useCallback, useState, useMemo } from "react";
import { getLocations, searchLocation } from "services/WeatherService";
import debounce from "lodash/debounce";
import styled from "styled-components";
import DatePicker from "react-datepicker";
import { useNavigate } from "react-router-dom";
import { addDays } from "date-fns";

import "react-datepicker/dist/react-datepicker.css";

const FlexBox = styled.div`
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  gap: 4em;
  padding: 2.8em 0;
`;

const InputField = styled.input`
  padding: 10px;
  border: 2px solid purple;
  color: purple;
  background-color: rgba(245, 245, 245, 0.2);
  border-radius: 4px;
  font-size: 16px;
  min-width: 50px;
  transition: border-color 0.3s;

  &:focus {
    border-color: #007bff;
    outline: none;
  }
  &:disabled {
    background-color: lightgrey;
    opacity: 1;
  }
`;
const Label = styled.label`
  width: 69px;
  font-size: 14px;
  font-weight: bold;
  text-align: left;
  padding: 10px;
  padding-left: 0px;
  color: purple;
  margin-bottom: 5px;
`;

const SearchContainer = styled.div`
  display: flex;
  justify-content: center;
  align-items: center;
  gap: 3vw;
`;

const SearchButton = styled.button`
  padding: 10px 20px;
  border: none;
  border-radius: 4px;
  background-color: purple;
  color: #fff;
  font-size: 16px;
  cursor: pointer;
  transition: background-color 0.3s;
  opacity: ${(props) => (props.disabled ? 0.5 : 1)};
  pointer-events: ${(props) => (props.disabled ? "none" : "auto")};

  &:hover {
    background-color: #0056b3;
  }
`;

const Dropdown = styled.ul`
  list-style: none;
  padding: 0;
  margin-top: 41px;
  border: 1px solid #ccc;
  border-radius: 4px;
  position: absolute;
  width: 190px;
  background-color: rgba(0, 0, 0, 0.6);
  max-height: 150px;
  overflow-y: auto;
  z-index: 1000;
`;

const DropdownItem = styled.li`
  padding: 10px;
  cursor: pointer;

  &:hover {
    background-color: rgba(245, 245, 245, 0.2);
  }
`;

// Search weather by location name or coordinates
export const LocationSearch = ({ setLocations, setError }) => {
  const [startDate, setStartDate] = useState("");
  const [timezone, setTimezone] = useState(null);
  const [locations, setSearchLocations] = useState([]);
  const [longitude, setLongitude] = useState("");
  const [latitude, setLatitude] = useState("");
  const [locationQuery, setLocationQuery] = useState("");
  const [selectedResult, setSelectedResult] = useState(null);

  const navigate = useNavigate();

  const debouncedSearchLocation = useCallback(
    debounce(searchLocation, 500),
    [],
  );
  const handleInputChange = (e) => {
    const locationValue = e.target.value;
    setLocationQuery(locationValue);
    setSelectedResult(null);
    debouncedSearchLocation(setSearchLocations, setError, locationValue);
  };

  const handleKeyDown = (e) => {
    if (e.key === "Backspace") {
      setLocationQuery("");
      setSelectedResult(null);
      setLongitude("");
      setLatitude("");
    }
  };

  const handleLongitudeChange = (e) => {
    if (e.target.value <= 180 && e.target.value >= -180) {
      setLongitude(e.target.value);
    }
  };

  const handleLatitudeChange = (e) => {
    if (e.target.value <= 90 && e.target.value >= -90) {
      setLatitude(e.target.value);
    }
  };

  const handleResultClick = (result) => {
    setLocationQuery(`${result.name}, ${result.country}`);
    setLatitude(result.latitude);
    setLongitude(result.longitude);
    setTimezone(result.timezone);
    setSelectedResult(result);
    setSearchLocations([]);
  };

  const handleSearch = async (setLocations) => {
    const [destination] = await getLocations(
      setLocations,
      setError,
      latitude,
      longitude,
      startDate,
      timezone,
    );
    if (destination) {
      return navigate(`/weather/${selectedResult?.name || "newLocation"}`, {
        state: { destination, selectedResult },
      });
    }
  };

  const isSearchButtonDisabled = useMemo(
    () =>
      !Boolean(startDate) ||
      (!Boolean(selectedResult) && !Boolean(longitude)) ||
      !Boolean(latitude),
    [startDate, selectedResult, longitude, latitude],
  );

  return (
    <FlexBox>
      <SearchContainer>
        <Label htmlFor="date">Date*:</Label>
        <div>
          <DatePicker
            selected={startDate}
            onChange={(date) => setStartDate(date)}
            placeholderText="Pick a Date"
            maxDate={addDays(new Date(), 13)}
            customInput={
              <InputField type="text" required value={locationQuery} />
            }
          />
        </div>
      </SearchContainer>
      <div>
        <SearchContainer>
          <Label htmlFor="location">Location:</Label>
          <div style={{ display: "flex", flexDirection: "column" }}>
            <InputField
              id="location"
              type="text"
              disabled={
                !Boolean(selectedResult) &&
                (Boolean(longitude) || Boolean(latitude))
              }
              value={locationQuery}
              onChange={handleInputChange}
              onKeyDown={handleKeyDown}
              placeholder="City, Town or Zip Code"
            />

            {locations?.length > 0 && (
              <Dropdown>
                {locations.map((location, index) => (
                  <DropdownItem
                    key={index}
                    onClick={() => handleResultClick(location)}
                  >
                    {location?.name}, {location?.country}
                  </DropdownItem>
                ))}
              </Dropdown>
            )}
          </div>
        </SearchContainer>
        <p style={{ fontWeight: "bold", padding: "1em 0" }}>OR</p>
        <SearchContainer>
          <Label htmlFor="latitude">
            Latitude: <span style={{ fontSize: "10px" }}>(-90 to 90)</span>
          </Label>

          <InputField
            id="latitude"
            type="number"
            min="-90"
            max="90"
            disabled={Boolean(selectedResult)}
            value={latitude}
            onChange={handleLatitudeChange}
            style={{ width: "100%" }}
            placeholder="-5"
          />

          <Label htmlFor="longitude">
            Longitude: <span style={{ fontSize: "10px" }}>(-180 to 180)</span>
          </Label>
          <InputField
            id="longitude"
            type="number"
            min="-180"
            max="180"
            value={longitude}
            disabled={Boolean(selectedResult)}
            onChange={handleLongitudeChange}
            style={{ width: "100%" }}
            placeholder="10"
          />
        </SearchContainer>
      </div>
      <div>
        <SearchButton
          onClick={() => handleSearch(setLocations)}
          disabled={isSearchButtonDisabled}
        >
          Search
        </SearchButton>
      </div>
    </FlexBox>
  );
};
