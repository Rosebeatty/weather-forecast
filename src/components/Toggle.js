import Toggle from "react-toggle";
import styled from "styled-components";

import "react-toggle/style.css";

const Label = styled.div`
  position: absolute;
  top: 3%;
  right: 5%;
`;
const Icon = styled.p`
  padding-top: 0.3em;
  color: white;
  margin: 0;
`;

// Toggle to switch temperature units (째F/째C)
export const ToggleSwitch = ({ unit, changeUnit }) => (
  <Label htmlFor="toggle">
    <Toggle
      id="toggle"
      data-testid="toggle"
      defaultChecked={unit}
      icons={{
        checked: <Icon>째C</Icon>,
        unchecked: <Icon>째F</Icon>,
      }}
      onChange={changeUnit}
      aria-label="No label"
    />
  </Label>
);
