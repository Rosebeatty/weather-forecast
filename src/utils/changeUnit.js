export const changeUnit = (unit, setUnit) => {
  setUnit(!unit);
  localStorage.setItem("toggle", JSON.stringify(unit));
};
