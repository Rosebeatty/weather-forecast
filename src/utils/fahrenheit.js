export const fahrenheit = (c) => {
  return Math.round((c * 9) / 5 + 32);
};
