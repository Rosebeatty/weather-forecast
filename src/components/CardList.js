import React from "react";
import { Link } from "react-router-dom";
import styled from "styled-components";
import { TailSpin } from "react-loader-spinner";
import { fahrenheit } from "utils/fahrenheit";

const Cards = styled.div`
  display: flex;
  width: 100%;
  justify-content: center;
  flex-wrap: wrap;
`;
const Card = styled.div`
  flex: 0 1 24%;
  min-width: 22vw;
  min-height: 22vh;
  display: flex;
  flex-direction: row;
  justify-content: space-between;
  align-items: center;
  margin: 0.7em 1em;
  background-color: rgba(245, 245, 245, 0.2);
  border-radius: 5px;
  padding: 0 3.5em;
  &:hover {
    background-color: rgba(245, 245, 245, 0.1);
  }
  @media only screen and (max-width: 720px) {
    min-width: 50vw;
    min-height: 22vh;
  }
`;
const Loading = styled.div`
  height: 25vh;
  display: flex;
  align-items: center;
`;

export const CardList = ({ locations, unit }) => (
  <Cards>
    {locations ? (
      locations.map((destination, i) => {
        const { current, current_units, latitude, longitude } = destination;
        return (
          <div id={`${latitude}&${longitude}`} key={i}>
            <Link
              key={i}
              to={{
                pathname: `weather/localWeather`,
              }}
              state={{ destination, isLocal: true }}
              style={{ textDecoration: "none", color: "purple" }}
            >
              <Card>
                <h3>Local Weather</h3>
                <p id="unit">
                  {unit
                    ? `${Math.round(current.temperature_2m)} ${current_units.temperature_2m}`
                    : `${fahrenheit(current.temperature_2m)}Â°F`}
                </p>
              </Card>
            </Link>
          </div>
        );
      })
    ) : (
      <Loading>
        <TailSpin
          visible={true}
          height="80"
          width="80"
          color="purple"
          ariaLabel="tail-spin-loading"
          radius="1"
        />
      </Loading>
    )}
  </Cards>
);
