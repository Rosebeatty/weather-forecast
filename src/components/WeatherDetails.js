import styled from "styled-components";
import { fahrenheit } from "utils/fahrenheit";
import { format } from "date-fns";

const Grid = styled.div`
  display: flex;
  justify-content: center;
  flex-wrap: wrap;
  color: purple;
  margin: 0 auto;
  max-width: 75vw;
  @media only screen and (max-width: 720px) {
  }
`;
const Box = styled.div`
  padding: 0.7em;
  width: 25vw;

  &.box-1 {
    border-right: 1px solid purple;
    border-bottom: 1px solid purple;
  }
  &.box-2 {
    border-bottom: 1px solid purple;
  }
  &.box-4 {
    border-left: 1px solid purple;
  }
`;
const Card = styled.div`
  min-height: 35vh;
  max-width: 60vw;
  display: flex;
  color: purple;
  flex-direction: column;
  justify-content: center;
  margin: 1.7em auto;
  text-align: left;
  align-items: center;
  background-color: rgba(245, 245, 245, 0.2);
  border-radius: 5px;

  @media only screen and (max-width: 720px) {
    min-height: 38vh;
    min-width: 75vw;
  }
`;
const MaxMin = styled.div`
  display: flex;
  flex-direction: row;
  justify-content: space-evenly;
  gap: 0.5em;
  width: 40vw;
`;
const Temp = styled.h1`
  font-size: 3.7em;
  margin: 0.2em 0;
`;

// Weather details display for chosen location
function WeatherDetails({ destination, selectedResult, unit, isLocal }) {
  const { current, current_units, daily, daily_units } = destination;
  const sunrise = format(new Date(daily.sunrise[0]), "HH:mm");
  const sunset = format(new Date(daily.sunset[0]), "HH:mm");

  return (
    <div>
      <h1>
        {(selectedResult &&
          `${selectedResult?.name}, ${selectedResult?.country}`) ||
          (isLocal === true && "Local Weather") ||
          `${destination.longitude}° long - ${destination.longitude}° lat`}
      </h1>
      <h2 style={{ fontSize: "16px" }}>
        {format(new Date(daily.time[0]), "EEEE LLLL do y")}
      </h2>
      <Card>
        <Temp>
          <p id="unit" data-testid="unit">
            {unit
              ? `${Math.round(current.temperature_2m)} ${current_units.temperature_2m}`
              : `${fahrenheit(current.temperature_2m)}°F`}
          </p>
        </Temp>
        <MaxMin>
          <p>
            Min:{" "}
            {unit
              ? `${Math.round(daily.temperature_2m_min)} ${daily_units.temperature_2m_min}`
              : `${fahrenheit(daily.temperature_2m_min)}°F`}
          </p>
          <p>
            Max:{" "}
            {unit
              ? `${Math.round(daily.temperature_2m_max)} ${daily_units.temperature_2m_max}`
              : `${fahrenheit(daily.temperature_2m_max)}°F`}
          </p>
        </MaxMin>
      </Card>
      <Grid>
        {selectedResult && (
          <>
            <Box className="box-1">
              <h4>Sunrise</h4>
              <p>{sunrise}</p>
            </Box>
            <Box className="box-2">
              <h4>Sunset</h4>
              <p>{sunset}</p>
            </Box>
          </>
        )}
        <Box className="box-1">
          <h4>Wind Speed</h4>
          <p>
            {current.wind_speed_10m} {current_units.wind_speed_10m}
          </p>
        </Box>
        <Box className="box-2">
          <h4>Percipitation</h4>
          <p>
            {current.precipitation} {current_units.precipitation}
          </p>
        </Box>
        <Box>
          <h4>Humidity</h4>
          <p>
            {current.relative_humidity_2m} {current_units.relative_humidity_2m}
          </p>
        </Box>
      </Grid>
    </div>
  );
}

export default WeatherDetails;
