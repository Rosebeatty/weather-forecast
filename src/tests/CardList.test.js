import React from "react";
import { render, screen, waitFor } from "@testing-library/react";
import { BrowserRouter as Router } from "react-router-dom";
import { CardList } from "components/CardList";
import { mockLocations } from "./mockData";
import { fahrenheit } from "utils/fahrenheit";

describe("<CardList />", () => {
  const mockProps = {
    locations: mockLocations,
    unit: false,
  };

  beforeEach(() => {
    render(
      <Router>
        <CardList {...mockProps} />
      </Router>,
    );
  });

  it("renders without crashing", () => {
    const { asFragment } = render(
      <Router>
        <CardList {...mockProps} />
      </Router>,
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("should render 1 location's cards/links", () => {
    const links = screen.getAllByRole("link");
    expect(links).toHaveLength(1);
  });

  it("rounds decimals to whole numbers", () => {
    const unitElement = screen.getByText(
      (content, element) => element.id === "unit",
    );
    expect(unitElement.textContent).not.toContain(".");
  });

  it("converts Celsius to Fahrenheit", () => {
    const unitElement = screen.getByText(
      (content, element) => element.id === "unit",
    );
    const tempC = unitElement.textContent.split("°")[0];
    const tempF = fahrenheit(tempC);
    expect(tempF).toEqual(136);
  });

  it("tests link to Location page has correct props", async () => {
    await waitFor(() => {
      const linkElement = screen.getByRole("link", { name: /Local/i });
      expect(linkElement).toHaveAttribute("href", "/weather/localWeather");
    });
  });

  it("displays temperature on the first card", async () => {
    await waitFor(() => {
      const linkElement = screen.getByRole("link", { name: /58°F/i });
      expect(linkElement).toBeInTheDocument();
    });
  });
});
