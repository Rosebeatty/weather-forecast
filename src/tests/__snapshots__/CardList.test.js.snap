// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CardList /> renders without crashing 1`] = `
<DocumentFragment>
  <div
    class="sc-gsFSjX bKuoSs"
  >
    <div
      id="46&9.18"
    >
      <a
        href="/weather/localWeather"
        style="text-decoration: none; color: purple;"
      >
        <div
          class="sc-kAycRU frRPlU"
        >
          <h3>
            Local Weather
          </h3>
          <p
            id="unit"
          >
            58Â°F
          </p>
        </div>
      </a>
    </div>
  </div>
</DocumentFragment>
`;
